### PR TITLE
image-slideshow: Add features to slideshow source

### PR DESCRIPTION
**Pull request by cg2121**
This adds the following features to the slideshow source:
-Manual mode: With hotkeys the user can change slides manually
-Loop: Add option to loop slideshow
-Hotkeys: Add play/pause, restart, stop, next slide, previous slide hotkeys
-Hide on end: Add option to hide slideshow when it ends
-Visibility behavior: Add pause on inactive, restart when activated, and always play

### CODE REVIEW

File: plugins/image-source/obs-slideshow.c [hunk 1]
```diff
@@ -343,21 +425,161 @@ static void ss_update(void *data, obs_data_t *settings)
...
 
 	obs_data_array_release(array);
 }
 
+static void next_slide(void *data)
+{
+	struct slideshow *ss = data;
+
+	if (++ss->cur_item >= (int)ss->files.num) {
+			ss->cur_item = 0;
```
Comment (RytoEX) [inline plugins/image-source/obs-slideshow.c]: Style:  This should be indented from the `if` statement with one tab instead of two (two tabs total instead of three).

File: plugins/image-source/obs-slideshow.c [hunk 2]
```diff
@@ -343,21 +425,161 @@ static void ss_update(void *data, obs_data_t *settings)
...
+		do_transition(ss, false);
 
 	obs_data_array_release(array);
 }
 
+static void next_slide(void *data)
+{
+	struct slideshow *ss = data;
+
+	if (++ss->cur_item >= (int)ss->files.num) {
```
Comment (RytoEX) [inline plugins/image-source/obs-slideshow.c]: Style:  This `if` statement does not need curly braces since its contents only use one line.

File: plugins/image-source/obs-slideshow.c [hunk 3]
```diff
@@ -343,21 +425,161 @@ static void ss_update(void *data, obs_data_t *settings)
...
+	}
+
+	do_transition(ss, false);
+}
+
+static void previous_slide(void *data)
+{
+	struct slideshow *ss = data;
+
+	if (--ss->cur_item <= 0) {
```
Comment (RytoEX) [inline plugins/image-source/obs-slideshow.c]: Style:  This `if` statement does not need curly braces since its contents only use one line.

File: plugins/image-source/obs-slideshow.c [hunk 4]
```diff
@@ -343,21 +425,161 @@ static void ss_update(void *data, obs_data_t *settings)
...
+
+static void play_pause_hotkey(void *data, obs_hotkey_id id,
+		obs_hotkey_t *hotkey, bool pressed)
+{
+	UNUSED_PARAMETER(id);
+	UNUSED_PARAMETER(hotkey);
+
+	struct slideshow *ss = data;
+
+	if (pressed) {
```
Comment (RytoEX) [inline plugins/image-source/obs-slideshow.c]: Style:  This `if` statement does not need curly braces since its contents only use one line.

File: plugins/image-source/obs-slideshow.c [hunk 5]
```diff
@@ -343,21 +425,161 @@ static void ss_update(void *data, obs_data_t *settings)
...
+
+static void restart_hotkey(void *data, obs_hotkey_id id,
+		obs_hotkey_t *hotkey, bool pressed)
+{
+	UNUSED_PARAMETER(id);
+	UNUSED_PARAMETER(hotkey);
+
+	struct slideshow *ss = data;
+
+	if (pressed) {
```
Comment (RytoEX) [inline plugins/image-source/obs-slideshow.c]: Style:  This `if` statement does not need curly braces since its contents only use one line.

File: plugins/image-source/obs-slideshow.c [hunk 6]
```diff
@@ -343,21 +425,161 @@ static void ss_update(void *data, obs_data_t *settings)
...
+
+static void stop_hotkey(void *data, obs_hotkey_id id,
+		obs_hotkey_t *hotkey, bool pressed)
+{
+	UNUSED_PARAMETER(id);
+	UNUSED_PARAMETER(hotkey);
+
+	struct slideshow *ss = data;
+
+	if (pressed) {
```
Comment (RytoEX) [inline plugins/image-source/obs-slideshow.c]: Style:  This `if` statement does not need curly braces since its contents only use one line.

File: plugins/image-source/obs-slideshow.c [hunk 7]
```diff
@@ -343,21 +425,161 @@ static void ss_update(void *data, obs_data_t *settings)
...
+{
+	UNUSED_PARAMETER(id);
+	UNUSED_PARAMETER(hotkey);
+
+	struct slideshow *ss = data;
+
+	if (!ss->manual)
+		return;
+
+	if (pressed) {
```
Comment (RytoEX) [inline plugins/image-source/obs-slideshow.c]: Style:  This `if` statement does not need curly braces since its contents only use one line.

File: plugins/image-source/obs-slideshow.c [hunk 8]
```diff
@@ -343,21 +425,161 @@ static void ss_update(void *data, obs_data_t *settings)
...
```
Comment (RytoEX) [inline plugins/image-source/obs-slideshow.c]: Style:  This `if` statement does not need curly braces since its contents only use one line.

File: plugins/image-source/obs-slideshow.c [hunk 9]
```diff
@@ -343,21 +425,161 @@ static void ss_update(void *data, obs_data_t *settings)
...
 	if (ss->files.num)
-		obs_transition_start(ss->transition, OBS_TRANSITION_MODE_AUTO,
-				ss->tr_speed,
-				ss->files.array[ss->cur_item].source);
+		do_transition(ss, false);
 
 	obs_data_array_release(array);
 }
 
+static void next_slide(void *data)
```
Comment (RenegadeTech) [inline plugins/image-source/obs-slideshow.c]: This function will fail if null is passed in.

File: plugins/image-source/obs-slideshow.c (hunk 2 above)
Comment (RenegadeTech) [inline plugins/image-source/obs-slideshow.c]: Is it safe to pre-increment the pointer here? Why is this needed / being done?

Comment (RenegadeTech) [inline plugins/image-source/obs-slideshow.c]: Why do you think this?

Comment (RytoEX) [inline plugins/image-source/obs-slideshow.c]: Because the [style guide](https://github.com/torvalds/linux/blob/master/Documentation/process/coding-style.rst) linked to from this project's [CONTRIBUTING document](https://github.com/jp9000/obs-studio/blob/master/CONTRIBUTING.rst) specifically says so:

> Do not unnecessarily use braces where a single statement will do.

Comment (RenegadeTech) [inline plugins/image-source/obs-slideshow.c]: Sure, okay.

File: plugins/image-source/obs-slideshow.c (hunk 9 above)
Comment (cg2121) [inline plugins/image-source/obs-slideshow.c]: It shouldn't. No other existing functions have null checks.

Comment (RenegadeTech) [inline plugins/image-source/obs-slideshow.c]: Doesn't mean it shouldn't be encourage.

File: plugins/image-source/obs-slideshow.c [hunk 10]
```diff
@@ -193,12 +254,31 @@ static void ss_update(void *data, obs_data_t *settings)
...
+		ss->behavior = BEHAVIOR_PAUSE_UNPAUSE;
+	} else if (astrcmpi(behavior, S_BEHAVIOR_ALWAYS_PLAY) == 0) {
+		ss->behavior = BEHAVIOR_ALWAYS_PLAY;
+	} else { /* S_BEHAVIOR_STOP_RESTART */
+		ss->behavior = BEHAVIOR_STOP_RESTART;
+	}
+
+	mode = obs_data_get_string(settings, S_MODE);
+
+	if (astrcmpi(mode, S_MODE_MANUAL) == 0)
```
Comment (RenegadeTech) [inline plugins/image-source/obs-slideshow.c]: Why not this as:
```c
ss->manual = astrcmpi(mode, S_MODE_MANUAL) == 0;
```

**Comment by SuslikV** [comment]
I can understand the "Play/Pause" button, but the other two ("Restart", "Stop") equals to Hide source via eye icon. Isn't it?

File: plugins/image-source/obs-slideshow.c [hunk 11]
```diff
@@ -44,31 +63,56 @@ struct image_file_data {
...
+	bool paused;
+	bool stop;
 	float slide_time;
 	uint32_t tr_speed;
 	const char *tr_name;
 	obs_source_t *transition;
 
 	float elapsed;
-	size_t cur_item;
+	int cur_item;
```
Comment (notr1ch) [inline plugins/image-source/obs-slideshow.c]: What's the reason behind this change? Seems to introduce a lot of unnecessary casting later on.

File: plugins/image-source/obs-slideshow.c (hunk 9 above)
Comment (notr1ch) [inline plugins/image-source/obs-slideshow.c]: The OBS API should never pass null to a source callback. Adding null checks would mask possible API errors that should be fixed if this ever occurs.

File: plugins/image-source/obs-slideshow.c (hunk 11 above)
Comment (cg2121) [inline plugins/image-source/obs-slideshow.c]: If a user is at slide 0 and goes to previous slide, they would be at index -1 (which then the index will change from -1 to files.num-1). Int can be a negative number and size_t can't.

Comment (notr1ch) [inline plugins/image-source/obs-slideshow.c]: Shouldn't it be set to the last slide in that case? How is a negative number any more useful?

Comment (cg2121) [inline plugins/image-source/obs-slideshow.c]: I fixed it now so it doesn't need a negative number.

File: plugins/image-source/obs-slideshow.c [hunk 12]
```diff
@@ -416,13 +689,11 @@ static void ss_video_tick(void *data, float seconds)
 			ss->cur_item = next;
 
 		} else if (++ss->cur_item >= ss->files.num) {
-			ss->cur_item = 0;
+				ss->cur_item = 0;
```
Comment (notr1ch) [inline plugins/image-source/obs-slideshow.c]: Unnecessary tab added here

File: plugins/image-source/obs-slideshow.c [hunk 13]
```diff
@@ -179,6 +223,23 @@ static bool valid_extension(const char *ext)
...
+		obs_transition_set(ss->transition,
+				ss->files.array[ss->cur_item].source);
+	} else if (!to_null && !ss->use_cut) {
+		obs_transition_start(ss->transition,
+				OBS_TRANSITION_MODE_AUTO,
+				ss->tr_speed,
+				ss->files.array[ss->cur_item].source);
+	} else if (to_null && !ss->use_cut) {
+		obs_transition_set(ss->transition, NULL);
+	}
```
Comment (RytoEX) [inline plugins/image-source/obs-slideshow.c]: Style:  Curly braces are not needed for this `if` statement, since the contents of each branch is a single statement.

File: plugins/image-source/obs-slideshow.c [hunk 14]
```diff
@@ -193,12 +254,28 @@ static void ss_update(void *data, obs_data_t *settings)
...
 
+	behavior = obs_data_get_string(settings, S_BEHAVIOR);
+
+	if (astrcmpi(behavior, S_BEHAVIOR_PAUSE_UNPAUSE) == 0) {
+		ss->behavior = BEHAVIOR_PAUSE_UNPAUSE;
+	} else if (astrcmpi(behavior, S_BEHAVIOR_ALWAYS_PLAY) == 0) {
+		ss->behavior = BEHAVIOR_ALWAYS_PLAY;
+	} else { /* S_BEHAVIOR_STOP_RESTART */
+		ss->behavior = BEHAVIOR_STOP_RESTART;
+	}
```
Comment (RytoEX) [inline plugins/image-source/obs-slideshow.c]: Style:  Curly braces are not needed for this `if` statement, since the contents of each branch is a single statement.

**Comment by jp9000** [comment]
Making a minor adjustment to the text for manual mode to specify to the user that they need to use hotkeys then merging.  Pretty much a perfect PR here though otherwise.
